smbclient/shutil.py

```diff
@@ -0,0 +1,306 @@
+
+from ntpath import join, basename, normcase, abspath, normpath, splitdrive
+
+import sys
+import stat as py_stat
+from os import error as os_error
+
+from smbclient._io import ioctl_request, SMBFileTransaction, SMBRawIO
+from smbclient._os import remove, rmdir, listdir, lstat, stat, utime, open_file, makedirs, readlink, symlink
+from smbclient.path import islink, isdir
+from smbprotocol.exceptions import SMBOSError
+from smbprotocol.ioctl import SMB2SrvCopyChunk, IOCTLFlags, SMB2SrvCopyChunkResponse, CtlCode, SMB2SrvCopyChunkCopy, \
+    SMB2SrvRequestResumeKey
+from smbprotocol.open import FilePipePrinterAccessMask, CreateOptions, ShareAccess
+
+CHUNK_SIZE = 1 * 1024 * 1024  # maximum chunksize allowed by smb
+
+
+class Error(EnvironmentError):
+    pass
+
+
+class SpecialFileError(EnvironmentError):
+    """Raised when trying to do a kind of operation (e.g. copying) which is
+    not supported on a special file (e.g. a named pipe)"""
+
+
+def rmtree(path, ignore_errors=False, onerror=None, **kwargs):
+    """Recursively delete a directory tree.
+
+    If ignore_errors is set, errors are ignored; otherwise, if onerror
+    is set, it is called to handle the error with arguments (func,
+    path, exc_info) where func is smbclient.listdir, smbclient.remove,
+    or smbclient.rmdir;
+    path is the argument to that function that caused it to fail; and
+    exc_info is a tuple returned by sys.exc_info().  If ignore_errors
+    is false and onerror is None, an exception is raised.
+
+    :param path: The path to remove.
+    :param ignore_errors: The ignore errors flag.
+    :param onerror: The callback executed on errors
+    :param kwargs: Common arguments used to build the SMB Session.
+    :return: True if path is a dir or points to a dir.
+    """
+    if ignore_errors:
+        def onerror(*args):
+            pass
+    elif onerror is None:
+        def onerror(*args):
+            raise
+    try:
+        if islink(path, **kwargs):
+            # symlinks to directories are forbidden, see bug #1669
+            raise OSError("Cannot call rmtree on a symbolic link")
+    except OSError:
+        onerror(islink, path, sys.exc_info())
+        # can't continue even if onerror hook returns
+        return
+
+    names = []
+    try:
+        names = listdir(path, **kwargs)
+    except os_error:
+        onerror(listdir, path, sys.exc_info())
+    for name in names:
+        fullname = join(path, name)
+        try:
+            mode = lstat(fullname, **kwargs).st_mode
+        except os_error:
+            mode = 0
+        if py_stat.S_ISDIR(mode):
+            rmtree(fullname, ignore_errors, onerror, **kwargs)
+        else:
+            try:
+                remove(fullname, **kwargs)
+            except os_error:
+                onerror(remove, fullname, sys.exc_info())
+    try:
+        rmdir(path, **kwargs)
+    except os_error:
+        onerror(rmdir, path, sys.exc_info())
+
+
+def copy2(src, dst, server_side_copy=False, **kwargs):
```

Comment:
So what we're interested in is mainly server side copy of files and entire directory trees. Local <-> remote was not considered at all, your're right. We could add this, by checking the UNC paths and if it's a UNC path at all. If I understand your suggestion correctly, you still wouldn't call that "copy2", because of more things listed above, that need to be taken into consideration?
Maybe adding that points to the copy2 documentation is an alternative? The python documentation does it that way for example (https://docs.python.org/3/library/shutil.html). I also have no problem in renaming the functions to reserve the name for a more complete one in the future.

Reply:
I think we should have `copy` at a minimum and potentially look at `copyfileobj` snf `copyfile`. The `copy2`, `copymode`, and `copystat` should be reserved for the future when we have a clearer picture and more time to really think about how to implement it here. Due to the complexity I think it should be done in a separate PR and not as part of this basic implementation.

As for the local <-> remote question, I'm happy to just have remote <-> remote in this PR for now but remove the `server_side_copy` option. We can still raise a ValueError if the path isn't remote and just remove that in a future iteration. I don't think it warrants a `copy2` consideration, it still fits under the `copy` spec where it's just copying the file's metadata, smbclient just allows it to be done over SMB :).

Reply:
OK got it. Just the point with server_side_copy is not clear to me. Let's talk just about remote copy for a moment: Here, we can only do server_side_copy, if the server supports it (e.g. recent SAMBA versions) and we're on the same share/server. If not, we have to fall back to reading and writing chunks from python. Not sure how to find out if the server supports it though. That's why I've introduced the flag to indicate, that the server supports it (default is False).

The python chunk copy (which requires up and download of the files) can be easily reinforced to also work with local files.

Reply:
Hmm ok I wasn't aware of that, looking at the Samba docs I can see that it introduced support for `FSCTL_SRV_COPYCHUNK` in 4.1.0 https://wiki.samba.org/index.php/Server-Side_Copy which has been EOL since 2016 https://wiki.samba.org/index.php/Samba_Release_Planning.

It's hard to tell when it was supported on the server side for Windows but considering

1. https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/eaa129d1-fca4-46b1-8ccd-b91f3e1d2d80#Appendix_A_Target_132 doesn't mention a minimum dialect version
2. https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb/c82ff91d-28d8-4274-a2ab-c2f5ade26b98 - It is supported in SMB1

I would say it has been supported on Windows for a while now.

Because of this I feel like we can rely on the fact that the server supports the server side copy operations and we don't need to fall back on our own manual read/write of chunks. If someone was using the ancient version of Samba I don't really want to make things easier for them (reason why I didn't add SMB1 support). As for whether it is still remote <-> remote but different server/share then we can just check the path to make sure they are the same before using the IOCTL commands or falling back to our own manual chunk read/writing.

Lastly, this isn't necessary but it sounds like the pure server/share copy through IOCTL should be a function in `smbclient._os` and `smbclient.shutil` just calls that if it's able to. The `._os` function is designed to expose the pure SMB side in an easy to use interface which this sounds like a good candidate. To better explain what I mean

```
# smbclient._os.py
def copy(src, dst, **kwargs):
    check that src and dst are on the same share
    Call IOCTL FSCTL_SRV_COPYCHUNK to copy the files


# smbclient.shutil.py
def copy(src, dst, **kwargs):
    if src and dst are on the same share
        call _os.py's copy(src, dst, **kwargs)
    else
        manually read and write from src/dst ourselves
```

Reply:
I have renamed it and removed all the functionality to copy between local and shares or between shares for now.